### PR TITLE
Implement BYOS plugin template and Slurm plugin example

### PR DESCRIPTION
**Pull request by demartinofra**
Signed-off-by: Francesco De Martino <fdm@amazon.com>

**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Available Jinja variables:
```
Comment (lukeseawalker) [inline scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml:1]: what about using for this file and the followings the paths
`/scheduler_plugins/slurm/` and `/scheduler_plugins/template/`?

In this way it's aligned with the definition of the template, e.g. 
```
ClusterInfrastructure:
  CloudFormation:
    Template: s3://<BUCKET>/scheduler_plugins/slurm/slurm_plugin_infrastructure.cfn.yaml
```

File: scheduler_plugins/slurm_plugin/artifacts/init_cookbook.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+#
+set -e
+
+tar -xf artifacts.tar.gz
```
Comment (lukeseawalker) [inline scheduler_plugins/slurm_plugin/artifacts/init_cookbook.sh]: isn't this done already in the HeadInit event?

File: scheduler_plugins/plugin_template/artifacts/handlers/compute_configure.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+# - PCLUSTER_EC2_INSTANCE_TYPE: type of the EC2 instance where the action is running
+# - PCLUSTER_OS: OS distribution of the instance where the script is running
+# - PCLUSTER_ARCH: architecture of the node: x86_64 or arm64
+# - PCLUSTER_VERSION: version of ParallelCluster the cluster belongs to.
+# - PCLUSTER_HEADNODE_PRIVATE_IP: private IP of the head node
+# - PCLUSTER_HEADNODE_HOSTNAME: hostname of the head node
+# - PCLUSTER_QUEUE_NAME: contains the name of the queue the node belongs to (empty in case of head node).
+# - PCLUSTER_COMPUTE_RESOURCE_NAME: contains the name of the compute resource the node belongs to (empty in case of head node).
+# - PCLUSTER_INSTANCE_TYPES_DATA: path to a JSON file containing the instance types data for the instances used in the queues
+# - PCLUSTER_NODE_TYPE: type of the node: HEAD or COMPUTE
```
Comment (lukeseawalker) [inline scheduler_plugins/plugin_template/artifacts/handlers/compute_configure.sh]: HEAD and COMPUTE lowercase, here and in the other files

File: scheduler_plugins/slurm_plugin/artifacts/handlers/head_finalize.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+set -e
+
+sudo -E cinc-client -z -o slurm_plugin_cookbook::head_node_finalize --log-level info --force-formatter
```
Comment (lukeseawalker) [inline scheduler_plugins/slurm_plugin/artifacts/handlers/head_finalize.sh]: for recipe names, what about using the convention <phase>-<node-type or name>, e.g.
finalize_head_node
init_compute
install_slurm
etc...

File: scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml (hunk 1 above)
Comment (demartinofra) [reply scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml:1]: I'd keep them in the root for now being them the main entrypoints of the plugin

Comment (demartinofra) [reply scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml:1]: ok I had misunderstood the comment. Renamed slurm_plugin root directory to slurm

File: scheduler_plugins/plugin_template/plugin_definition.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  # Defaults to all ParallelCluster version compatible with the interface version declared in PluginInterfaceVersion
+  # SupportedParallelClusterVersions: ">=3.0.0"
+# Define the additional AWS resources that need to be created/updated/deleted together with a cluster.
+ClusterInfrastructure:
+  CloudFormation:
+    # Reference to a CloudFormation template that is created as a nested stack of the main cluster stack.
+    # In order to allow access to regionalized resources the ${Region} and the ${URLSuffix} variables are made available
+    # when defining this url. The template can be specified as an S3 or HTTPS url.
+    Template: s3://<BUCKET>/scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml
+    # Template: https://<BUCKET>.s3.${AWS::Region}.${AWS::URLSuffix}/scheduler_plugins/plugin_template/additional_cluster_infrastructure.cfn.yaml
```
Comment (lukeseawalker) [inline scheduler_plugins/plugin_template/plugin_definition.yaml]: ${AWS::Region} -> ${Region}
and ${AWS::URLSuffix} -> ${URLSuffix}

File: scheduler_plugins/plugin_template/plugin_definition.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+PluginResources:
+  # HTTPS or S3 URL of remote resources that ParallelCluster downloads and makes available to plugin event handlers.
+  # In order to allow access to regionalized resources the ${Region} and the ${URLSuffix} variables are made available
+  # when defining this String.
+  ClusterSharedArtifacts:
+    - Source: s3://<BUCKET>/scheduler_plugins/plugin_template/artifacts.tar.gz
+      # Source: https://<BUCKET>.s3.${AWS::Region}.${AWS::URLSuffix}/scheduler_plugins/plugin_template/artifacts.tar.gz
```
Comment (lukeseawalker) [inline scheduler_plugins/plugin_template/plugin_definition.yaml]: same here

**Review by lukeseawalker — APPROVED** [review]

**Pull request merged**